src/main/java/bio/terra/tanagra/query/Query.java

```diff
@@ -0,0 +1,167 @@
+import java.util.List;
+import java.util.stream.Collectors;
+import org.stringtemplate.v4.ST;
+
+public record Query(
+    List<FieldVariable> select,
+    List<TableVariable> tables,
+    FilterVariable where,
+    List<OrderByVariable> orderBy,
+    List<FieldVariable> groupBy,
+    HavingFilterVariable having,
+    Integer limit)
+    implements SQLExpression {
+
+  public Query {
+    if (select.isEmpty()) {
+      throw new IllegalArgumentException("Query must have at least one SELECT FieldVariable");
+    }
+    if (tables.isEmpty()) {
+      throw new IllegalArgumentException("Query must have at least one TableVariable");
+    }
+    if (groupBy == null) {
+      groupBy = List.of();
+    }
+    if (orderBy == null) {
+      orderBy = List.of();
+    }
+    if (!groupBy.isEmpty() && !orderBy.isEmpty()) {
+      throw new IllegalArgumentException("Query cannot have both GROUP BY and ORDER BY");
+    }
+    long primaryTables = tables.stream().filter(TableVariable::isPrimary).count();
+    if (primaryTables != 1) {
+      throw new IllegalArgumentException(
+          "Query can only have one primary table, but found " + primaryTables);
+    }
+  }
+
+  public Query(List<FieldVariable> select, List<TableVariable> tables) {
+    this(select, tables, null, null, null, null, null);
+  }
+
+  public Query(
+      List<FieldVariable> select, List<TableVariable> tables, List<FieldVariable> groupBy) {
+    this(select, tables, null, null, groupBy, null, null);
+  }
+
+  public Query(List<FieldVariable> select, List<TableVariable> tables, FilterVariable where) {
+    this(select, tables, where, null, null, null, null);
+  }
+
+  public Query(
+      List<FieldVariable> select,
+      List<TableVariable> tables,
+      List<OrderByVariable> orderBy,
+      Integer limit) {
+    this(select, tables, null, orderBy, null, null, limit);
+  }
+
+  public Query(
+      List<FieldVariable> select,
+      List<TableVariable> tables,
+      List<OrderByVariable> orderBy,
+      List<FieldVariable> groupBy,
+      Integer limit) {
+    this(select, tables, null, orderBy, groupBy, null, limit);
+  }
+
+  public Query(List<FieldVariable> select, List<TableVariable> tables, int limit) {
+    this(select, tables, null, null, null, null, limit);
+  }
+
+  public static final String TANAGRA_FIELD_PREFIX = "t_";
+
+  @Override
+  public String renderSQL(SqlPlatform platform) {
+    // generate a unique alias for each TableVariable
+    TableVariable.generateAliases(tables);
+
+    // render each SELECT FieldVariable and join them into a single string
+    String selectSQL =
+        select.stream()
+            .sorted(Comparator.comparing(FieldVariable::getAlias))
+            .map(fieldVariable -> fieldVariable.renderSQL(platform))
+            .collect(Collectors.joining(", "));
+
+    if (platform == SqlPlatform.SYNAPSE && limit != null) {
+      selectSQL = "TOP " + limit + " " + selectSQL;
+    }
+
+    // render the primary TableVariable
+    String sql =
```

Comment:
I'm not sure how that would work, but I'd much rather use the prepared statement API if possible. I think you need a SQL connection to create one, so it would be hard to do in the current flow; that would require passing in some state in addition to `platform`.